enhancements/machine-api/control-plane-machine-set.md

```diff
@@ -0,0 +1,908 @@
+MachineSets have a single `ProviderSpec` which defines the configuration for the Machines created by the MachineSet.
+The failure domain (sometimes called availability zone) is a part of this provider spec and as such, defines that all
+Machines within the Machineset share a failure domain.
+
+This is undesirable for Control Plane Machines as we wish to have them spread across multiple availability zones to
+reduce the likelihood of datacenter level faults degrading the control plane.
+To this end, the failure domains for the control plane will be set on the `ControlPlaneMachineSet` spec directly.
+
+When creating Machines, the `ControlPlaneMachineSet` controller will balance the Machines across these failure domains
+by injecting the desired failure domain for the new Machine into the provider spec based on the platform specific
+failure domain field.
+
+##### Failures domains
+
+The `FailureDomains` field is expected to be populated by the user/installer based on the topology of the Control Plane
+Machines. For example, we expect on AWS that this will contain a list of availability zones within a single region.
+Note, we are explicitly not expecting users to add different regions to the the `FailureDomains` as we do not support
+running OpenShift across multiple regions.
+
+The following table denotes the field on each platform into which the list of failure domains will be mapped:
+
+
+| Platform  | ProviderSpec Field          |
+| --------- | --------------------------- |
+| AWS       | .placement.availabilityZone |
+| Azure     | .zone                       |
+| GCP       | .zone                       |
+| vSphere   | TBD (see below)             |
+| OpenStack | .availabilityZone           |
+
+Note that on some platforms (eg OpenStack) the failure domain field is optional, as such, the `FailureDomains` field
+must also be optional.
+
+###### Failure Domains on vSphere
+
+As of OpenShift 4.10, there is no concept of a zone within vSphere and as such we would expect that the `FailureDomains`
+would be omitted for vSphere.
+
+However, [there is future work](https://github.com/openshift/enhancements/pull/918) to include zone support for vSphere
+within 4.11. Once this is implemented, the Machine API provider for vSphere will understand the concept of zones,
+defined within the `Infrastructure` resource.
+The `ControlPlaneMachineSet` will then be able to balance Machines across the different zones by setting the new zone
+information within the provider specs as it does on other platforms.
+
+#### Ensuring Machines match the desired state
+
+To ensure Machines are up to date with the desired configuration, the `ControlPlaneMachineSet` controller will leverage
+a similar pattern to the workload controllers.
+It will hash the template and compare the hashed result with the spec of the Machines within the managed set.
+As we expect the failure domain to vary across the managed Machines, this will need to be omitted before the hash can
```

Comment:
There are more fields than just the failure domain that will vary across the managed Machines. For example, on AWS, the `Subnet` field will vary based upon which failure domain is chosen.

Reply:
We had a few discussions about this and how different platforms handle this kind of difference, I think AWS didn't even cross our minds, but the likes of vSphere and OpenStack etc did.

I will need to revisit this topic and see what we can come up with. There has been talk of `FailureDomain` CRDs (CAPI do this for vSphere) that define all of the context required to create a Machine within a failure domain, the alternative is that we do something inline, for example

```
failureDomains:
- name: my-made-up-fd-1
  value:
    availabilityZone: us-east-1a
    subnet: my-subnet-1a
- name: my-made-up-fd-2
  value:
    availabilityZone: us-east-1b
    subnet: my-subnet-1b
```
With which we then overlay the information onto the shared providerSpec.

The other alternative is that they provide 1 providerSpec per failure domain but that isn't great from a UX perspective.

Reply:
An alternative here, which would be more concrete, would to be to create a discriminated union and have each of the fields that will be able to be overriden within the providerSpec as a concrete field, I think from an API perspective that is likely to be the safer option, this would probably look something like this instead
```
failureDomains:
- aws:
    availabilityZone: us-east-1a
    subnet: my-subnet-1a
- aws:
    availabilityZone: us-east-1b
    subnet: my-subnet-1b
```

Or for Azure:

```
failureDomains:
- azure:
    zone: us-central-1
- azure:
    zone: us-central-2
```

Reply:
I've pushed an update which redefines the failure domain as a discriminated union, which I think should solve the problem